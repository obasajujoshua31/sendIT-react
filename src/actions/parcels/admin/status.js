import axios from "axios";

import {
  signInFailure,
  signInSuccess,
  updateParcelStatus
} from "../../../types/types";
import { setSignInState } from "../../setState";

import { baseUrl } from "../../../helpers/constants";

const changeStatus = (parcelId, status) => dispatch => {
  const token = window.localStorage.getItem("token");
  return axios
    .put(
      `${baseUrl}/parcels/${parcelId}/status`,
      {
        status
      },
      {
        headers: {
          Authorization: token
        }
      }
    )
    .then(() => {
      dispatch(setSignInState(signInSuccess, null, true, null));
      dispatch({
        type: updateParcelStatus
      });
    })
    .catch(error => {
      dispatch(
        setSignInState(signInFailure, null, null, error.response.data.error)
      );
    });
};

export default changeStatus;
