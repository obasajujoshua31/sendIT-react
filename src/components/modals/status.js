import React, { Component } from "react";
import ReactModal from "react-modal";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import updateStatus from "../../actions/parcels/admin/status";

export class ChangeStatus extends Component {
  constructor(props) {
    super(props);
    this.state = {
      parcel: [],
      showCancelModal: false,
      status: ""
    };
    this.changeStatus = this.changeStatus.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  changeStatus(e) {
    e.preventDefault();
    this.props.changeParcelStatus(this.props.parcelId, this.state.status);
  }
  componentWillMount() {
    if (process.env.NODE_ENV !== "test") {
      ReactModal.setAppElement("#app");
    }
  }
  handleChange(e) {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  componentDidMount = () => {
    this.setState({ status: this.props.status });
  };

  componentDidUpdate(prevProps) {
    if (this.props !== prevProps) {
      if (this.props.parcelStatus) {
        this.props.history.push("/admin-dashboard");
      }
    }
  }

  render() {
    const { show, closeCancel } = this.props;

    return (
      <ReactModal
        isOpen={show}
        contentLabel="Modal #1 Global Style Override Example"
        onRequestClose={closeCancel}
        overlayClassName="Overlay"
        className="editModal"
      >
        <div style={modalHeaderStyle}>
          <span style={spanStyle}>
            <i className="far fa-times-circle" onClick={closeCancel} />
          </span>
          <h3 style={modalHeadBrandStyle}>Change Parcel Status</h3>
        </div>
        <div style={modalBodyContainer}>
          <form onSubmit={this.changeStatus} noValidate>
            <select
              style={selectStyle}
              onChange={this.handleChange}
              name="status"
              value={this.state.status}
            >
              <option value="CANCELLED" disabled>
                Cancelled
              </option>
              <option value="PLACED" disabled>
                Placed
              </option>
              <option value="TRANSITING">Transiting</option>
              <option value="DELIVERED">Delivered</option>
            </select>
            <br />
            <button style={buttonStyle}>Confirm Change Status</button>
          </form>
        </div>
      </ReactModal>
    );
  }
}

export const mapStateToProps = ({ parcels }) => ({
  parcelStatus: parcels.parcelStatus
});
export const mapDispatchToProps = dispatch => ({
  changeParcelStatus: (id, status) => dispatch(updateStatus(id, status))
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(ChangeStatus));

const spanStyle = {
  fontSize: "1.8em",
  color: "white",
  cursor: "pointer",
  padding: 5,
  position: "absolute",
  right: 5,
  bottom: 5
};

const modalHeaderStyle = {
  padding: 15,
  backgroundColor: "#806B75",
  position: "relative",
  display: "flex",
  justifyContent: "center",
  alignItem: "center"
};

const modalHeadBrandStyle = {
  fontWeight: "bold",
  fontSize: "1.4rem",
  color: "white"
};

const modalBodyContainer = {
  padding: "10px",
  maxWidth: "100%",
  display: "flex",
  justifyContent: "center",
  alignItem: "center"
};
const buttonStyle = {
  position: "relative",
  margin: 0,
  marginTop: "10px",
  cursor: "pointer",
  padding: "10px 20px",
  backgroundColor: "#fd9b01",
  color: "white",
  border: "none",
  outline: "none"
};
const selectStyle = {
  height: "30px",
  width: "200px"
};
