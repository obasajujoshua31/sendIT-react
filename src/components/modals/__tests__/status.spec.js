import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import { mapDispatchToProps, mapStateToProps, ChangeStatus } from "../status";

const props = {
  parcelId: "1",
  role: false,
  history: {
    push: jest.fn()
  },
  show: true,
  status: "PLACED",
  closeCancel: false,
  changeParcelStatus: jest.fn(),
  destination: "dkkdkd",
  parcelStatus: false
};
const mockStore = {
  parcels: {
    parcelStatus: true
  }
};
const mockSubmit = {
  preventDefault: jest.fn()
};

const mockHandleChange = value => ({
  target: {
    value,
    name: "status"
  }
});

const wrapper = shallow(<ChangeStatus {...props} />);

describe("Edit ChangeStatus Test", () => {
  it("should render Edit ChangeStatus", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("dispatch should call loadOneParcel", () => {
    const dispatchSpy = sinon.spy();
    const dispatchResult = mapDispatchToProps(dispatchSpy);
    dispatchResult.changeParcelStatus("1", "TRANSITING");
    expect(dispatchSpy.callCount).toEqual(1);
  });

  it("map state to props should return parcels", () => {
    const mapStateResult = mapStateToProps(mockStore);
    expect(mapStateResult.parcelStatus).toEqual(true);
  });

  it("changeStatus should call props.changeParcelStatus", () => {
    const changeParcelSpy = jest.spyOn(wrapper.instance(), "changeStatus");
    changeParcelSpy(mockSubmit);
    expect(wrapper.instance().props.changeParcelStatus).toBeCalledWith(
      "1",
      "PLACED"
    );
  });

  it("handle change should set errors when e.target.value is not defined", () => {
    const handleChangeSpy = jest.spyOn(wrapper.instance(), "handleChange");

    handleChangeSpy(mockHandleChange("TRANSITING"));
    expect(wrapper.instance().state.status).toEqual("TRANSITING");
  });

  it("componentDidMount should set state to this.prop.status", () => {
    const componentSpy = jest.spyOn(wrapper.instance(), "componentDidMount");

    wrapper.setProps({
      status: "TRANSITING"
    });
    componentSpy();
    expect(wrapper.instance().state.status).toEqual("TRANSITING");
  });

  it("should call props.history.push when parcelStatus is true", () => {
    wrapper.setProps({
      parcelStatus: true
    });
    expect(wrapper.instance().props.history.push).toBeCalledWith(
      "/admin-dashboard"
    );
  });

  it("should render status modal if state is the same as props", () => {
    wrapper.setState({
      parcelStatus: false
    });
    wrapper.setProps({
      parcelStatus: false
    });
    expect(wrapper).toMatchSnapshot();
  });
});
