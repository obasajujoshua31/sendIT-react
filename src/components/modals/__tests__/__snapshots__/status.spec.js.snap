// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit ChangeStatus Test should render Edit ChangeStatus 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="editModal"
  closeTimeoutMS={0}
  contentLabel="Modal #1 Global Style Override Example"
  isOpen={true}
  onRequestClose={false}
  overlayClassName="Overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    style={
      Object {
        "alignItem": "center",
        "backgroundColor": "#806B75",
        "display": "flex",
        "justifyContent": "center",
        "padding": 15,
        "position": "relative",
      }
    }
  >
    <span
      style={
        Object {
          "bottom": 5,
          "color": "white",
          "cursor": "pointer",
          "fontSize": "1.8em",
          "padding": 5,
          "position": "absolute",
          "right": 5,
        }
      }
    >
      <i
        className="far fa-times-circle"
        onClick={false}
      />
    </span>
    <h3
      style={
        Object {
          "color": "white",
          "fontSize": "1.4rem",
          "fontWeight": "bold",
        }
      }
    >
      Change Parcel Status
    </h3>
  </div>
  <div
    style={
      Object {
        "alignItem": "center",
        "display": "flex",
        "justifyContent": "center",
        "maxWidth": "100%",
        "padding": "10px",
      }
    }
  >
    <form
      noValidate={true}
      onSubmit={[Function]}
    >
      <select
        name="status"
        onChange={[Function]}
        style={
          Object {
            "height": "30px",
            "width": "200px",
          }
        }
        value="PLACED"
      >
        <option
          disabled={true}
          value="CANCELLED"
        >
          Cancelled
        </option>
        <option
          disabled={true}
          value="PLACED"
        >
          Placed
        </option>
        <option
          value="TRANSITING"
        >
          Transiting
        </option>
        <option
          value="DELIVERED"
        >
          Delivered
        </option>
      </select>
      <br />
      <button
        style={
          Object {
            "backgroundColor": "#fd9b01",
            "border": "none",
            "color": "white",
            "cursor": "pointer",
            "margin": 0,
            "marginTop": "10px",
            "outline": "none",
            "padding": "10px 20px",
            "position": "relative",
          }
        }
      >
        Confirm Change Status
      </button>
    </form>
  </div>
</Modal>
`;

exports[`Edit ChangeStatus Test should render status modal if state is the same as props 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="editModal"
  closeTimeoutMS={0}
  contentLabel="Modal #1 Global Style Override Example"
  isOpen={true}
  onRequestClose={false}
  overlayClassName="Overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    style={
      Object {
        "alignItem": "center",
        "backgroundColor": "#806B75",
        "display": "flex",
        "justifyContent": "center",
        "padding": 15,
        "position": "relative",
      }
    }
  >
    <span
      style={
        Object {
          "bottom": 5,
          "color": "white",
          "cursor": "pointer",
          "fontSize": "1.8em",
          "padding": 5,
          "position": "absolute",
          "right": 5,
        }
      }
    >
      <i
        className="far fa-times-circle"
        onClick={false}
      />
    </span>
    <h3
      style={
        Object {
          "color": "white",
          "fontSize": "1.4rem",
          "fontWeight": "bold",
        }
      }
    >
      Change Parcel Status
    </h3>
  </div>
  <div
    style={
      Object {
        "alignItem": "center",
        "display": "flex",
        "justifyContent": "center",
        "maxWidth": "100%",
        "padding": "10px",
      }
    }
  >
    <form
      noValidate={true}
      onSubmit={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "preventDefault": [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    >
      <select
        name="status"
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "target": Object {
                    "name": "status",
                    "value": "TRANSITING",
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        style={
          Object {
            "height": "30px",
            "width": "200px",
          }
        }
        value="TRANSITING"
      >
        <option
          disabled={true}
          value="CANCELLED"
        >
          Cancelled
        </option>
        <option
          disabled={true}
          value="PLACED"
        >
          Placed
        </option>
        <option
          value="TRANSITING"
        >
          Transiting
        </option>
        <option
          value="DELIVERED"
        >
          Delivered
        </option>
      </select>
      <br />
      <button
        style={
          Object {
            "backgroundColor": "#fd9b01",
            "border": "none",
            "color": "white",
            "cursor": "pointer",
            "margin": 0,
            "marginTop": "10px",
            "outline": "none",
            "padding": "10px 20px",
            "position": "relative",
          }
        }
      >
        Confirm Change Status
      </button>
    </form>
  </div>
</Modal>
`;
